plugins/inputs/docker_log/docker_logs.go

```diff
@@ -0,0 +1,440 @@
+				return 0, er
+			}
+		}
+		stream := StdType(buf[stdWriterFdIndex])
+		// Check the first byte to know where to write
+		var logType string
+		switch stream {
+		case Stdin:
+			logType = "stdin"
+			break
+		case Stdout:
+			logType = "stdout"
+			break
+		case Stderr:
+			logType = "stderr"
+			break
+		case Systemerr:
+			fallthrough
+		default:
+			return 0, fmt.Errorf("Unrecognized input header: %d", buf[stdWriterFdIndex])
+		}
+		// Retrieve the size of the frame
+		frameSize = int(binary.BigEndian.Uint32(buf[stdWriterSizeIndex : stdWriterSizeIndex+4]))
+
+		// Check if the buffer is big enough to read the frame.
+		// Extend it if necessary.
+		if frameSize+stdWriterPrefixLen > bufLen {
+			buf = append(buf, make([]byte, frameSize+stdWriterPrefixLen-bufLen+1)...)
+			bufLen = len(buf)
+		}
+
+		// While the amount of bytes read is less than the size of the frame + header, we keep reading
+		for nr < frameSize+stdWriterPrefixLen {
+			var nr2 int
+			nr2, er = src.Read(buf[nr:])
+			nr += nr2
+			if er == io.EOF {
+				if nr < frameSize+stdWriterPrefixLen {
+					return written, nil
+				}
+				break
+			}
+			if er != nil {
+				return 0, er
+			}
+		}
+
+		// we might have an error from the source mixed up in our multiplexed
+		// stream. if we do, return it.
+		if stream == Systemerr {
+			return written, fmt.Errorf("error from daemon in stream: %s", string(buf[stdWriterPrefixLen:frameSize+stdWriterPrefixLen]))
+		}
+
+		tags["logType"] = logType
+		fields := map[string]interface{}{}
+		fields["log"] = buf[stdWriterPrefixLen+1 : frameSize+stdWriterPrefixLen]
+		acc.AddFields("docker_log", fields, tags)
+		written += int64(frameSize)
+
+		// Move the rest of the buffer to the beginning
+		copy(buf, buf[frameSize+stdWriterPrefixLen:])
+		// Move the index
+		nr -= frameSize + stdWriterPrefixLen
+	}
+}
+
+func (d *DockerLogs) Start(acc telegraf.Accumulator) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	var c Client
+	var err error
+	if d.Endpoint == "ENV" {
+		c, err = d.newEnvClient()
+	} else {
+		tlsConfig, err := d.ClientConfig.TLSConfig()
+		if err != nil {
+			return err
+		}
+		c, err = d.newClient(d.Endpoint, tlsConfig)
+	}
+	if err != nil {
+		return err
+	}
+	d.client = c
+	// Create label filters if not already created
+	if !d.filtersCreated {
+		err := d.createLabelFilters()
+		if err != nil {
+			return err
+		}
+		err = d.createContainerFilters()
+		if err != nil {
+			return err
+		}
+		err = d.createContainerStateFilters()
+		if err != nil {
+			return err
+		}
+		d.filtersCreated = true
+	}
+	filterArgs := filters.NewArgs()
+	for _, state := range containerStates {
+		if d.stateFilter.Match(state) {
+			filterArgs.Add("status", state)
+		}
+	}
+
+	// All container states were excluded
+	if filterArgs.Len() == 0 {
+		return nil
+	}
+
+	d.opts = types.ContainerListOptions{
+		Filters: filterArgs,
+	}
+	return nil
+}
+
+func (d *DockerLogs) Stop() {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	d.stopAllReaders()
+	d.wg.Wait()
```

Comment:
I believe we can deadlock here because we don't release the lock until all the goroutines complete.  Once we add the required locking as suggested in my other comments, the goroutines will need to wait for this lock so they can remove themselves from the containerList.

Fix is easy though, just release the lock after calling stopAllReaders.

Reply:
Moved d.mu.Unlock to after stopAllReaders()